homeassistant/components/decora_wifi/__init__.py

```diff
@@ -1 +1,135 @@
-"""The decora_wifi component."""
+"""Support for the myLeviton decora_wifi component."""
+
+from datetime import timedelta
+import logging
+
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import (
+    CONF_ENTITY_ID,
+    CONF_HOST,
+    CONF_ID,
+    CONF_PASSWORD,
+    CONF_SCAN_INTERVAL,
+    CONF_USERNAME,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryNotReady
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity_component import EntityComponent
+from homeassistant.helpers.entity_platform import EntityPlatform
+
+from .common import CommFailed, DecoraWifiPlatform, LoginFailed, LoginMismatch
+from .const import CONF_OPTIONS, CONF_TITLE, DEFAULT_SCAN_INTERVAL, DOMAIN, PLATFORMS
+
+_LOGGER = logging.getLogger(__name__)
+
+DECORAWIFI_HOST_SCHEMA = vol.Schema({vol.Required(CONF_HOST): cv.string})
+
+CONFIG_SCHEMA = vol.Schema(
+    vol.All(
+        cv.deprecated(DOMAIN),
+        {
+            DOMAIN: vol.Schema(
+                {
+                    vol.Required(CONF_USERNAME): cv.string,
+                    vol.Required(CONF_PASSWORD): cv.string,
+                    vol.Optional(CONF_SCAN_INTERVAL, default=120): cv.positive_int,
+                },
+            )
+        },
+    ),
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+async def async_setup(hass: HomeAssistant, config):
+    """Set up the Decora Wifi component."""
+    hass.data.setdefault(
+        DOMAIN, EntityComponent(_LOGGER, DOMAIN, hass, timedelta(DEFAULT_SCAN_INTERVAL))
+    )
+
+    if DOMAIN not in config:
+        return True
+
+    # Trigger import config flow
+    conf = config[DOMAIN]
+
+    if not hass.config_entries.async_entries(DOMAIN):
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": config_entries.SOURCE_IMPORT},
+                data={
+                    CONF_USERNAME: conf[CONF_USERNAME],
+                    CONF_PASSWORD: conf[CONF_PASSWORD],
+                },
+            )
+        )
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry):
+    """Set up Decora Wifi from a config entry."""
+
+    conf_data = dict(entry.data)
+    email = conf_data[CONF_USERNAME]
+    password = conf_data[CONF_PASSWORD]
+
+    component: EntityComponent = hass.data[DOMAIN]
```

Comment:
Don't use EntityComponent here, its not intended for use with integrations setup by config entries.

Reply:
The upstream requires a session object instance to manage its connection. As I explained in my previous responses, this session object can't be stored in config data. I reiterate here: even if the session object provided by the upstream was JSON serializable--**which it is not**--the result of retrieving it would be a separate clone of the object, **not** the same object. 

The obvious pythonic solution is to store the object in a dict and use lifecycle events to remove it from the dict when the config entry is removed from hass, which is what I had done previously. Your previous comment rejected that approach, and this is the workaround.  If neither approach is acceptable, please say so now, because if that's the case we're wasting our time here.

Reply:
Ok, I can give that a shot this evening.  I guess I assumed that the hass backend would be expecting an EntityComponent in `hass.data[DOMAIN]` and if I tried to store a dict of objects in there I'd be overwriting it.

Reply:
The integration is solely responsible for `hass.data[DOMAIN]` so no need to worry about it being modified externally